src/OpenTelemetry.Contrib.PersistentStorage/PersistentStorageHelper.cs

```diff
@@ -0,0 +1,260 @@
+                        success = true;
+                    }
+                    catch (Exception ex)
+                    {
+                        StorageEventSource.Log.Warning("File rename of {filePath} to {newFilePath} has failed.", ex);
+                    }
+                }
+            }
+
+            return success;
+        }
+
+        internal static bool RemoveTimedOutTmpFiles(DateTime timeoutDeadline, string filePath)
+        {
+            bool success = false;
+
+            if (filePath.EndsWith(".tmp", StringComparison.OrdinalIgnoreCase))
+            {
+                DateTime fileDateTime = GetDateTimeFromBlobName(filePath);
+                if (fileDateTime < timeoutDeadline)
+                {
+                    try
+                    {
+                        File.Delete(filePath);
+                        success = true;
+                        StorageEventSource.Log.Warning("File write exceeded timeout. Dropping telemetry");
+                    }
+                    catch (Exception ex)
+                    {
+                        StorageEventSource.Log.Warning($"Deletion of file {filePath} has failed.", ex);
+                    }
+                }
+            }
+
+            return success;
+        }
+
+        internal static void RemoveExpiredBlobs(string directoryPath, long retentionPeriodInMilliseconds, long writeTimeoutInMilliseconds)
+        {
+            var currentUtcDateTime = DateTime.UtcNow;
+
+            var leaseDeadline = currentUtcDateTime;
+            var retentionDeadline = currentUtcDateTime - TimeSpan.FromMilliseconds(retentionPeriodInMilliseconds);
+            var timeoutDeadline = currentUtcDateTime - TimeSpan.FromMilliseconds(writeTimeoutInMilliseconds);
+
+            foreach (var file in Directory.EnumerateFiles(directoryPath).OrderByDescending(f => f))
+            {
+                var success = RemoveTimedOutTmpFiles(timeoutDeadline, file);
+
+                if (success)
+                {
+                    continue;
+                }
+
+                success = RemoveExpiredLease(leaseDeadline, file);
+
+                if (!success)
+                {
+                    RemoveExpiredBlob(retentionDeadline, file);
+                }
+            }
+
+            // It is faster to calculate the directory size, instead of removing length of expired files.
+            var size = CalculateFolderSize(directoryPath);
+            Interlocked.Exchange(ref directorySize, size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static long GetDirectorySize()
+        {
+            return Interlocked.Read(ref directorySize);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void WriteAllBytes(string path, byte[] buffer)
+        {
+            File.WriteAllBytes(path, buffer);
+            UpdateDirectorySize(buffer.Length);
```

Comment:
What if the file got removed by other processes, do we update/fix the directory size periodically in the maintenance job?

Reply:
We do not check the directorySize periodically. We only update the directory size when a new file is added or a file is removed. The folder where data is stored is unique to one process so chances of other process removing the file is less assuming each process has unique identity.
```c#
string appIdentity = Environment.UserName + "@" + Path.Combine(baseDirectory, Process.GetCurrentProcess().ProcessName);
string subdirectoryName = GetSHA256Hash(appIdentity);
subdirectoryPath = Path.Combine(path, subdirectoryName);
Directory.CreateDirectory(subdirectoryPath);
```

Reply:
> The folder where data is stored is unique to one process so chances of other process removing the file is less assuming each process has unique identity (not guaranteed).

I'm bit confused - what is the definition of "one process"? Is it a unique instance of process (identified by PID)? If that's the case, a process restart (which would have a different PID) would cause the new process not able to pick up the file left by the previous process.

Reply:
My current understanding is that we affinitize the storage folder to an "app", and by default the app is identified by user name + proc name. Let's say we have `dotnet.exe` CLI being invoked thousands of times during a project build time, each of them would have different PID, but we treat them as the same app, so they could run concurrently and access the same folder.

Reply:
This is my bad
1) We do recalculate the directorySize on maintenance job.
2) And Yes storage folder is affinitize to an "app"